File: sdks/go/primitives/u256.go
package primitives

import (
	"encoding/hex"
	"fmt"
	"math/big"
	"strings"
)

// U256 represents a 256-bit unsigned integer
type U256 struct {
	bytes [32]byte // Little-endian representation
}

// NewU256 creates a new U256 from a uint64
func NewU256(value uint64) U256 {
	var u U256
	// Store in little-endian format
	for i := 0; i < 8; i++ {
		u.bytes[i] = byte(value >> (i * 8))
	}
	return u
}

// U256FromBytes creates a U256 from a byte slice (big-endian)
func U256FromBytes(b []byte) (U256, error) {
	if len(b) > 32 {
		return U256{}, fmt.Errorf("invalid U256 length: expected at most 32 bytes, got %d", len(b))
	}
	
	var u U256
	// Convert from big-endian to little-endian
	for i, j := 0, len(b)-1; j >= 0; i, j = i+1, j-1 {
		u.bytes[i] = b[j]
	}
	return u, nil
}

// U256FromHex creates a U256 from a hex string
func U256FromHex(s string) (U256, error) {
	// Remove 0x prefix if present
	s = strings.TrimPrefix(s, "0x")
	
	// Pad with leading zeros if necessary
	if len(s)%2 != 0 {
		s = "0" + s
	}
	
	if len(s) > 64 {
		return U256{}, fmt.Errorf("invalid hex U256 length: expected at most 64 characters, got %d", len(s))
	}
	
	bytes, err := hex.DecodeString(s)
	if err != nil {
		return U256{}, fmt.Errorf("invalid hex U256: %w", err)
	}
	
	return U256FromBytes(bytes)
}

// U256FromBigInt creates a U256 from a big.Int
func U256FromBigInt(b *big.Int) (U256, error) {
	if b.Sign() < 0 {
		return U256{}, fmt.Errorf("U256 cannot be negative")
	}
	
	// Check if the number fits in 256 bits
	if b.BitLen() > 256 {
		return U256{}, fmt.Errorf("number too large for U256")
	}
	
	bytes := b.Bytes() // big-endian
	return U256FromBytes(bytes)
}

// ZeroU256 returns a U256 with value 0
func ZeroU256() U256 {
	return U256{}
}

// OneU256 returns a U256 with value 1
func OneU256() U256 {
	return NewU256(1)
}

// MaxU256 returns the maximum U256 value (2^256 - 1)
func MaxU256() U256 {
	var u U256
	for i := range u.bytes {
		u.bytes[i] = 0xFF
	}
	return u
}

// Bytes returns the U256 as a byte slice (big-endian)
func (u U256) Bytes() []byte {
	result := make([]byte, 32)
	// Convert from little-endian to big-endian
	for i, j := 0, 31; i < 32; i, j = i+1, j-1 {
		result[j] = u.bytes[i]
	}
	return result
}

// Array returns the U256 as a byte array (little-endian internal representation)
func (u U256) Array() [32]byte {
	return u.bytes
}

// Hex returns the U256 as a hex string with 0x prefix
func (u U256) Hex() string {
	bytes := u.Bytes() // big-endian
	// Trim leading zeros
	start := 0
	for start < len(bytes) && bytes[start] == 0 {
		start++
	}
	
	if start == len(bytes) {
		return "0x0"
	}
	
	return "0x" + hex.EncodeToString(bytes[start:])
}

// String returns the U256 as a decimal string
func (u U256) String() string {
	return u.ToBigInt().String()
}

// ToBigInt converts the U256 to a big.Int
func (u U256) ToBigInt() *big.Int {
	bytes := u.Bytes() // big-endian
	return new(big.Int).SetBytes(bytes)
}

// Uint64 returns the U256 as a uint64 (panics if too large)
func (u U256) Uint64() uint64 {
	// Check if the value fits in uint64
	for i := 8; i < 32; i++ {
		if u.bytes[i] != 0 {
			panic("U256 value too large for uint64")
		}
	}
	
	var result uint64
	for i := 0; i < 8; i++ {
		result |= uint64(u.bytes[i]) << (i * 8)
	}
	return result
}

// TryUint64 returns the U256 as a uint64 and a boolean indicating success
func (u U256) TryUint64() (uint64, bool) {
	// Check if the value fits in uint64
	for i := 8; i < 32; i++ {
		if u.bytes[i] != 0 {
			return 0, false
		}
	}
	
	var result uint64
	for i := 0; i < 8; i++ {
		result |= uint64(u.bytes[i]) << (i * 8)
	}
	return result, true
}

// IsZero returns true if the U256 is zero
func (u U256) IsZero() bool {
	for _, b := range u.bytes {
		if b != 0 {
			return false
		}
	}
	return true
}

// Equal returns true if two U256s are equal
func (u U256) Equal(other U256) bool {
	return u.bytes == other.bytes
}

// Less returns true if u < other
func (u U256) Less(other U256) bool {
	// Compare from most significant byte (index 31) to least significant (index 0)
	for i := 31; i >= 0; i-- {
		if u.bytes[i] < other.bytes[i] {
			return true
		}
		if u.bytes[i] > other.bytes[i] {
			return false
		}
	}
	return false // Equal
}

// Greater returns true if u > other
func (u U256) Greater(other U256) bool {
	return other.Less(u)
}

// LessOrEqual returns true if u <= other
func (u U256) LessOrEqual(other U256) bool {
	return !u.Greater(other)
}

// GreaterOrEqual returns true if u >= other
func (u U256) GreaterOrEqual(other U256) bool {
	return !u.Less(other)
}

// Add performs addition with overflow detection
func (u U256) Add(other U256) (U256, error) {
	var result U256
	var carry uint16
	
	for i := 0; i < 32; i++ {
		sum := uint16(u.bytes[i]) + uint16(other.bytes[i]) + carry
		result.bytes[i] = byte(sum)
		carry = sum >> 8
	}
	
	if carry != 0 {
		return U256{}, fmt.Errorf("addition overflow")
	}
	
	return result, nil
}

// Sub performs subtraction with underflow detection
func (u U256) Sub(other U256) (U256, error) {
	if u.Less(other) {
		return U256{}, fmt.Errorf("subtraction underflow")
	}
	
	var result U256
	var borrow uint16
	
	for i := 0; i < 32; i++ {
		diff := uint16(u.bytes[i]) - uint16(other.bytes[i]) - borrow
		if diff > 255 { // Underflow occurred
			result.bytes[i] = byte(diff + 256)
			borrow = 1
		} else {
			result.bytes[i] = byte(diff)
			borrow = 0
		}
	}
	
	return result, nil
}

// Mul performs multiplication with overflow detection
func (u U256) Mul(other U256) (U256, error) {
	// Use big.Int for multiplication to detect overflow
	a := u.ToBigInt()
	b := other.ToBigInt()
	result := new(big.Int).Mul(a, b)
	
	// Check if result fits in 256 bits
	if result.BitLen() > 256 {
		return U256{}, fmt.Errorf("multiplication overflow")
	}
	
	return U256FromBigInt(result)
}

// Div performs division with division by zero detection
func (u U256) Div(other U256) (U256, error) {
	if other.IsZero() {
		return U256{}, fmt.Errorf("division by zero")
	}
	
	a := u.ToBigInt()
	b := other.ToBigInt()
	result := new(big.Int).Div(a, b)
	
	return U256FromBigInt(result)
}

// Mod performs modulo operation
func (u U256) Mod(other U256) (U256, error) {
	if other.IsZero() {
		return U256{}, fmt.Errorf("modulo by zero")
	}
	
	a := u.ToBigInt()
	b := other.ToBigInt()
	result := new(big.Int).Mod(a, b)
	
	return U256FromBigInt(result)
}

// MarshalText implements encoding.TextMarshaler
func (u U256) MarshalText() ([]byte, error) {
	return []byte(u.Hex()), nil
}

// UnmarshalText implements encoding.TextUnmarshaler
func (u *U256) UnmarshalText(text []byte) error {
	val, err := U256FromHex(string(text))
	if err != nil {
		return err
	}
	*u = val
	return nil
}

File: sdks/typescript/src/primitives/u256.ts
/**
 * Represents a 256-bit unsigned integer
 */
export class U256 {
  private readonly _value: bigint;

  private constructor(value: bigint) {
    if (value < 0n) {
      throw new Error('U256 cannot be negative');
    }
    if (value >= 2n ** 256n) {
      throw new Error('Value too large for U256');
    }
    this._value = value;
  }

  /**
   * Creates a U256 from a regular number
   */
  static fromNumber(value: number): U256 {
    if (!Number.isInteger(value) || value < 0) {
      throw new Error('Value must be a non-negative integer');
    }
    return new U256(BigInt(value));
  }

  /**
   * Creates a U256 from a bigint
   */
  static fromBigInt(value: bigint): U256 {
    return new U256(value);
  }

  /**
   * Creates a U256 from a hex string
   */
  static fromHex(hex: string): U256 {
    // Remove 0x prefix if present
    const cleanHex = hex.startsWith('0x') ? hex.slice(2) : hex;
    
    if (cleanHex.length === 0) {
      return U256.zero();
    }
    
    if (cleanHex.length > 64) {
      throw new Error(`Hex string too long for U256: ${cleanHex.length} characters`);
    }
    
    if (!/^[0-9a-fA-F]+$/.test(cleanHex)) {
      throw new Error('Invalid hex string: contains non-hex characters');
    }
    
    return new U256(BigInt('0x' + cleanHex));
  }

  /**
   * Creates a U256 from bytes (big-endian)
   */
  static fromBytes(bytes: Uint8Array | ArrayLike<number>): U256 {
    const byteArray = new Uint8Array(bytes);
    
    if (byteArray.length > 32) {
      throw new Error(`Too many bytes for U256: ${byteArray.length}`);
    }
    
    let value = 0n;
    for (let i = 0; i < byteArray.length; i++) {
      value = (value << 8n) + BigInt(byteArray[i] ?? 0);
    }
    
    return new U256(value);
  }

  /**
   * Creates a U256 with value 0
   */
  static zero(): U256 {
    return new U256(0n);
  }

  /**
   * Creates a U256 with value 1
   */
  static one(): U256 {
    return new U256(1n);
  }

  /**
   * Creates the maximum U256 value (2^256 - 1)
   */
  static max(): U256 {
    return new U256(2n ** 256n - 1n);
  }

  /**
   * Convenience method to create U256 from various inputs
   * @param input - Number, bigint, hex string, or byte array
   */
  static from(input: number | bigint | string | Uint8Array | ArrayLike<number>): U256 {
    if (typeof input === 'number') {
      return U256.fromNumber(input);
    }
    if (typeof input === 'bigint') {
      return U256.fromBigInt(input);
    }
    if (typeof input === 'string') {
      return U256.fromHex(input);
    }
    return U256.fromBytes(input);
  }

  /**
   * Returns the value as a bigint
   */
  toBigInt(): bigint {
    return this._value;
  }

  /**
   * Returns the value as a regular number (throws if too large)
   */
  toNumber(): number {
    if (this._value > BigInt(Number.MAX_SAFE_INTEGER)) {
      throw new Error('Value too large to convert to number');
    }
    return Number(this._value);
  }

  /**
   * Tries to convert to a regular number, returns null if too large
   */
  tryToNumber(): number | null {
    if (this._value > BigInt(Number.MAX_SAFE_INTEGER)) {
      return null;
    }
    return Number(this._value);
  }

  /**
   * Returns the value as a hex string with 0x prefix
   */
  toHex(): string {
    if (this._value === 0n) {
      return '0x0';
    }
    return `0x${this._value.toString(16)}`;
  }

  /**
   * Returns the value as a hex string with 0x prefix, padded to 64 characters
   */
  toHexPadded(): string {
    return `0x${this._value.toString(16).padStart(64, '0')}`;
  }

  /**
   * Returns the value as bytes (big-endian, 32 bytes)
   */
  toBytes(): Uint8Array {
    const bytes = new Uint8Array(32);
    let value = this._value;
    
    for (let i = 31; i >= 0; i--) {
      bytes[i] = Number(value & 0xFFn);
      value >>= 8n;
    }
    
    return bytes;
  }

  /**
   * Returns the value as a decimal string
   */
  toString(): string {
    return this._value.toString();
  }

  /**
   * Returns true if the value is zero
   */
  isZero(): boolean {
    return this._value === 0n;
  }

  /**
   * Returns true if two U256s are equal
   */
  equals(other: U256): boolean {
    return this._value === other._value;
  }

  /**
   * Returns true if this < other
   */
  lt(other: U256): boolean {
    return this._value < other._value;
  }

  /**
   * Returns true if this <= other
   */
  lte(other: U256): boolean {
    return this._value <= other._value;
  }

  /**
   * Returns true if this > other
   */
  gt(other: U256): boolean {
    return this._value > other._value;
  }

  /**
   * Returns true if this >= other
   */
  gte(other: U256): boolean {
    return this._value >= other._value;
  }

  /**
   * Addition with overflow checking
   */
  add(other: U256): U256 {
    const result = this._value + other._value;
    if (result >= 2n ** 256n) {
      throw new Error('Addition overflow');
    }
    return new U256(result);
  }

  /**
   * Subtraction with underflow checking
   */
  sub(other: U256): U256 {
    if (this._value < other._value) {
      throw new Error('Subtraction underflow');
    }
    return new U256(this._value - other._value);
  }

  /**
   * Multiplication with overflow checking
   */
  mul(other: U256): U256 {
    const result = this._value * other._value;
    if (result >= 2n ** 256n) {
      throw new Error('Multiplication overflow');
    }
    return new U256(result);
  }

  /**
   * Division with division by zero checking
   */
  div(other: U256): U256 {
    if (other._value === 0n) {
      throw new Error('division by zero');
    }
    return new U256(this._value / other._value);
  }

  /**
   * Modulo operation
   */
  mod(other: U256): U256 {
    if (other._value === 0n) {
      throw new Error('modulo by zero');
    }
    return new U256(this._value % other._value);
  }

  /**
   * Bitwise AND
   */
  and(other: U256): U256 {
    return new U256(this._value & other._value);
  }

  /**
   * Bitwise OR
   */
  or(other: U256): U256 {
    return new U256(this._value | other._value);
  }

  /**
   * Bitwise XOR
   */
  xor(other: U256): U256 {
    return new U256(this._value ^ other._value);
  }

  /**
   * Bitwise NOT
   */
  not(): U256 {
    return new U256((2n ** 256n - 1n) ^ this._value);
  }

  /**
   * Left shift
   */
  shl(bits: number): U256 {
    if (bits < 0 || bits >= 256) {
      throw new Error('Shift amount out of range');
    }
    const result = this._value << BigInt(bits);
    if (result >= 2n ** 256n) {
      throw new Error('Shift overflow');
    }
    return new U256(result);
  }

  /**
   * Right shift
   */
  shr(bits: number): U256 {
    if (bits < 0 || bits >= 256) {
      throw new Error('Shift amount out of range');
    }
    return new U256(this._value >> BigInt(bits));
  }

  /**
   * JSON serialization
   */
  toJSON(): string {
    return this.toHex();
  }

  /**
   * Create from JSON
   */
  static fromJSON(json: string): U256 {
    return U256.fromHex(json);
  }
}